File: framework/test2/struct/0359/lograte.rb
# step [1]: implement solution function
class Logger
  def initialize()
    @msgMap = {}
    @Limit = 10
    @lastTs = -@Limit
  end

=begin
  :type timestamp: Integer
  :type message: String
  :rtype: Boolean
=end
  def should_print_message(timestamp, message)
    _T = @lastTs
    @lastTs = timestamp
    if timestamp - _T >= @Limit
      @msgMap = {}
      @msgMap[message] = timestamp
      return true
    end
    if timestamp - @msgMap.fetch(message, timestamp - @Limit) < @Limit
      return false
    end
    @msgMap[message] = timestamp
    true
  end

end

if __FILE__ == $0
  require 'soda/unittest/work'
  ns = Soda::Unittest
  # step [2]: setup function/return/arguments
  # work = ns::TestWork.create(method(:my_solution))
  work = ns::TestWork.forStruct(Logger)
  # work.validator = -> (e, r) { e is equal to r }
  work.compareSerial = true
  puts work.run(ARGF.read)
end

File: framework/test2/struct/0359/lograte.ts
import {
    AvlTree,
    BinarySearchTree
} from '@datastructures-js/binary-search-tree';
import {
    PriorityQueue,
    MinPriorityQueue,
    MaxPriorityQueue
} from '@datastructures-js/priority-queue';
import { ListNode, NestedInteger } from 'soda/leetcode';
import { TestWork, Utils, Validators } from 'soda/unittest';

// step [1]: implement solution function
class Logger {
    private msgMap: Map<string, number>;
    private readonly Limit: number = 10;
    private lastTs: number;

    constructor() {
        this.msgMap = new Map<string, number>();
        this.lastTs = -this.Limit;
    }

    /** 
     * @param {number} timestamp 
     * @param {string} message
     * @return {boolean}
     */
    shouldPrintMessage(timestamp: number, message: string): boolean {
        let T = this.lastTs;
        this.lastTs = timestamp;
        if (timestamp - T >= this.Limit) {
            this.msgMap = new Map();
            this.msgMap.set(message, timestamp);
            return true;
        }

        let v = this.msgMap.get(message);
        if (typeof v === 'undefined') {
            v = timestamp - this.Limit;
        }
        if (timestamp - v < this.Limit) {
            return false;
        }
        this.msgMap.set(message, timestamp);
        return true;
    }
}

// step [2]: setup function/return/arguments
// let taskFunc = add;
// const work = TestWork.create<ReturnType<typeof taskFunc>>(taskFunc);
let work = TestWork.forStruct(Logger);
// work.validator = (x, y) => { ... };
work.compareSerial = true;
console.log(work.run(Utils.fromStdin()));
